### PR TITLE
added experiments

### PR DESCRIPTION
**Pull request by l-htfld**
Полное собрание экспериментов с корпусами

### CODE REVIEW

**Comment by fortvivlan** [comment]
во-первых, я что-то не поняла, вы используете bert base uncased?? но "Pretrained model on **English** language"
во-вторых, не до конца понимаю, что происходит в тетрадках с семантикой: часть до sarcasm_model.train() включительно мне понятна, почему вы далее опять обучаете исходную модельку, не понимаю. 
Кажется, не очень технически разумно каждый раз в getitem запускать модель для разметки семантики.

**Comment by fortvivlan** [comment]
1. В момент, когда вы инициализируете модель, она инициируется рандомными весами. И эти веса на каждом запуске разные, если не:
```
def set_random_seed(seed):
    torch.backends.cudnn.deterministic = True
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    np.random.seed(seed)
    random.seed(seed)
```
Эта функция - копипаста из домашнего задания номер 2. Вы как будто не делали его...
2. А оценить качество классификатора семантики не? Если у него качество близкое к нулю? Тогда и ваши метрики итогового классификатора сарказма, конечно, лучше точно не станут. Вообще при таком раскладе уж лучше извлечь из обученного классификатора семантики потюненные веса берта (SemanticModel.bert) и заменить ими стандартный берт в базовой архитектуре, заморозив их.

**Comment by fortvivlan** [comment]
```
parts = line.split('\t')
                if parts[0].isdigit() or '-' in parts[0]:
                    word = parts[1]
                    sem_class = parts[target_column] if len(parts) > 10 else 'O'
```
Такой вопрос - вы же CoBaLD Rus брали? https://github.com/CobaldAnnotation/CobaldRus Там не может быть токенов с дефисом в нумерации. Это возможно только в английском датасете, где есть мультитокены (don't, won't и подобные). Соответственно, и вероятность выполнения условия `len(parts) > 10` равна нулю (она в принципе равна нулю). Если первый символ в строке - это цифра, то это 100% токен и 100% для него заполнены все 12 полей. Для чего тогда отсеиваете? 

`pooled = logits.squeeze(0).mean(dim=0).cpu().numpy()`
Вопрос, насколько полезно добавлять сырые логиты семантической модели в классификатор сарказма, спорный: они совершенно не интерпретируемы. С другой стороны, ванхоты и в2в тоже не особо интерпретируемы, так что способ как способ. Просто про это я бы в тексте написала как минимум абзац: о способах добавления семантической информации для задачи текстовой классификации. Впрочем, это в основном Ксюшин диплом, не ваш. 

```
for param in sarcasm_model.bert.parameters():
    param.requires_grad = False
```
У меня нет уверенности, что вы понимаете, зачем я предложила это сделать. Заморозка параметров исходного берта означает, что веса берта не тюнятся под нашу задачу, обучается только голова классификатора и слой семантики. В данной версии решения морозить берт не особо осмысленно, следует тогда уж и с незамороженным обучить - чтобы оценить разницу. 

```
# Создаём и обучаем SarcasmClassifier
bert = sem_model.bert
sarcasm_model = SarcasmClassifier(
    bert_model=bert,
    sem_dim=df['sem_pooled'].iloc[0].shape[0],
    num_classes=2
).to(device)

for param in sarcasm_model.bert.parameters():
    param.requires_grad = False

#optimizer = AdamW(sarcasm_model.parameters(), lr=2e-5)

optimizer = AdamW(
    filter(lambda p: p.requires_grad, sarcasm_model.parameters()),
    lr=2e-5
)

sarcasm_model.train()
for epoch in range(3):
    total_loss = 0.0
    for batch in dl:
        optimizer.zero_grad()
        out = sarcasm_model(
            input_ids=batch['input_ids'].to(device),
            attention_mask=batch['attention_mask'].to(device),
            sem_feats=batch['sem_feats'].to(device),
            labels=batch['labels'].to(device)
        )
        out['loss'].backward()
        optimizer.step()
        total_loss += out['loss'].item()
    print(f"Sarcasm Epoch {epoch+1}, Loss={total_loss/len(dl):.4f}")

# Сохраняем модель
torch.save(sarcasm_model.state_dict(), 'sarcasm_model.pt')
print("Training complete.")
```
Этот кусок кода у вас какую роль выполняет? По-моему, его надо удалить. 

Так и не увидела, где метрики обеих моделей, SemanticModel и SarcasmClassifier.
